FiniteStateMachine/src/statemachine/StateMachine.java

```diff
@@ -3,18 +3,27 @@
 import org.jetbrains.annotations.NotNull;
 import statemachine.exceptions.BadTransitionException;
 
-import java.util.EnumSet;
+import java.util.Set;
 
+import static statemachine.exceptions.BadTransitionException.badTransitionException;
 
-public class StateContainer<E extends Enum<E> & State<E>> implements State<E> {
+
+/**
+ * A machine that has states of type {@code E} and knows its current state.
+ * It can transition between states, and will throw a {@link BadTransitionException}
+ * if an invalid transition is attempted.
+ */
+public class StateMachine<E extends State<E>> {
```

Comment:
It doesn't make sense to ask an individual `State` for its own state – it's always going to be `this`, by definition. It _does_ make sense to ask the state _machine_ what its current state is. But that's a concern of the machine, not the state. The only method they both have in common is `getStates`, but just because they both have that method doesn't mean they should share a common supertype. In fact, `State#getStates` should probably be renamed `getNextStates` and `StateMachine#getStates` should probably be renamed to `getNextStatesForCurrentState` or something. It's not obvious that `StateMachine#getStates` doesn't return all the possible states the machine could have, for example.